boofuzz/primitives/string.py

```diff
@@ -258,61 +248,32 @@ def add_long_strings(self, sequence):
         for string in strings:
             self._fuzz_library.append(string)
 
-    def mutate(self):
+    def mutations(self):
         """
         Mutate the primitive by stepping through the fuzz library extended with the "this" library, return False on
         completion.
 
         @rtype:  bool
         @return: True on success, False otherwise.
         """
+        for val in self._fuzz_library + self.this_library:
+            if self.size < 0 or len(val) <= self.size:
+                yield val
 
-        # loop through the fuzz library until a suitable match is found.
-        while 1:
-            # if we've ran out of mutations, raise the completion flag.
-            if self._mutant_index == self.num_mutations():
-                self._fuzz_complete = True
-
-            # if fuzzing was disabled or complete, and mutate() is called, ensure the original value is restored.
-            if not self._fuzzable or self._fuzz_complete:
-                self._value = self._original_value
-                return False
-
-            # update the current value from the fuzz library.
-            self._value = (self._fuzz_library + self.this_library)[self._mutant_index]
-
-            # increment the mutation count.
-            self._mutant_index += 1
-
-            # if the size parameter is disabled, done.
-            if self.size == -1:
-                return True
-
-            # ignore library items greater then user-supplied length.
-            # TODO: might want to make this smarter.
-            if len(self._value) > self.size:
-                continue
-            else:
-                return True
+    def encode(self, value, mutation_context):
+        if isinstance(value, six.text_type):
+            value = helpers.str_to_bytes(value)
+        # pad undersized library items.
+        if len(value) < self.size:
+            value += self.padding * (self.size - len(value))
+        return helpers.str_to_bytes(value)
 
-    def num_mutations(self):
+    def num_mutations(self, default_value):
         """
         Calculate and return the total number of mutations for this individual primitive.
 
         @rtype:  int
         @return: Number of mutated forms this primitive can take
+        :param default_value:
         """
         return len(self._fuzz_library) + len(self.this_library)
```

Comment:
As to the `default_value` parameter, it's not used yet. According to the comments in `Fuzzable` class, it's used if exhausting the mutations() is not possible. So maybe check if it's not `None`, then just return it as the number of mutations.

Reply:
The default_value parameter isn't used in this case because it doesn't impact the number of mutations for a String. Although it could in the future.

See bytes.py for an example of num_mutations that uses the default value.

But here in string.py, it needs to be here to match the method signature provided by Fuzzable.

Reply:
Got it, thanks. I regarded the `default_value` as the default number of mutations, actually it's the default value of the `fuzz_object`. Since this attribute is seperated from fuzz_object itself, there is no explict way to get it from the `fuzz_object`.